src/kernel/sporadic.c

```diff
@@ -210,135 +214,84 @@ void refill_update(sched_context_t *sc, ticks_t new_period, ticks_t new_budget,
-         * rAmount in a refill must be >= MIN_BUDGET */
+    if (!refill_single(sc)) {
+        ticks_t amount = refill_head(sc)->rAmount;
+        ticks_t tail = refill_head(sc)->rTime + amount;
+        return refill_index(sc, refill_next(sc, sc->scRefillHead))->rTime <= tail;
+    } else {
+        return false;
     }
 }
 
 void refill_budget_check(ticks_t usage)
 {
     sched_context_t *sc = NODE_STATE(ksCurSC);
-    /* this function should only be called when the sc is out of budget */
-    ticks_t capacity = refill_capacity(NODE_STATE(ksCurSC), usage);
-    assert(capacity < MIN_BUDGET || refill_full(sc));
-    assert(sc->scPeriod > 0);
+    assert(!isRoundRobin(sc));
     REFILL_SANITY_START(sc);
 
-    if (capacity == 0) {
-        while (refill_head(sc)->rAmount <= usage) {
-            /* exhaust and schedule replenishment */
-            usage -= refill_head(sc)->rAmount;
-            if (refill_single(sc)) {
-                /* update in place */
-                refill_head(sc)->rTime += sc->scPeriod;
-            } else {
-                refill_t old_head = refill_pop_head(sc);
-                old_head.rTime = old_head.rTime + sc->scPeriod;
-                schedule_used(sc, old_head);
-            }
-        }
-
-        /* budget overrun */
-        if (usage > 0) {
-            /* budget reduced when calculating capacity */
-            /* due to overrun delay next replenishment */
-            refill_head(sc)->rTime += usage;
-            /* merge front two replenishments if times overlap */
-            if (!refill_single(sc) &&
-                refill_head(sc)->rTime + refill_head(sc)->rAmount >=
-                refill_index(sc, refill_next(sc, sc->scRefillHead))->rTime) {
-
-                refill_t refill = refill_pop_head(sc);
-                refill_head(sc)->rAmount += refill.rAmount;
-                refill_head(sc)->rTime = refill.rTime;
-            }
-        }
+    /* We assume that the estimates on worst-case kernel overheads are
+     * correct and that overrun cannot occur. */
+#ifdef CONFIG_DEBUG_BUILD
+    if (usage > refill_head(sc)->rAmount) {
+        ticks_t overrun = usage - refill_head(sc)->rAmount;
+        userError(
+            "SC (%p) overran budget by %lluus (%llu ticks)",
+            sc,
+            ticksToUs(overrun),
+            overrun
+        );
     }
-
-    capacity = refill_capacity(sc, usage);
-    if (capacity > 0 && refill_ready(sc)) {
-        refill_split_check(usage);
+#endif
+    usage = MIN(usage, refill_head(sc)->rAmount);
+
+    /* Charge the usage to the head refill */
+    if (usage > 0) {
+        refill_t used = (refill_t) {
+            .rAmount = usage,
+            .rTime = refill_head(sc)->rTime + sc->scPeriod,
+        };
+
+        refill_head(sc)->rAmount -= usage;
+        /* We need to keep the head refill no more than a period before
+         * the start of the tail refill. */
+        refill_head(sc)->rTime += usage;
```

Comment:
Our proofs contain an invariant that the start time of a  tail refill in the refill list is no more than 1 period after the start time of the head refill. Whilst this is not itself relevant for real-time theory, it is used as part of the proof that the refills in the queue are always in chronological order, i.e. that a refill appended to the queue one period after its being made available will be no earlier than the current tail.

What this actually seems to achieve is the stronger outcome that the end time of the newly added tail is no less than one period after the start time of the head, which I suspect means that that is what is specifically proven here and is only later weakened to show that it satisfies the invariant.

Reply:
Since I don't know the proof side, there's not much I can add to this. My only suggestion would be to explain this in the comment

I presume the proof shows that this doesn't interfere with the usual scheduling guarantees provided by sporadic servers? It seems plausible to me, but not entirely obvious. Would be nice to record this claim in a comment.

Reply:
The proof so far doesn't provide any real scheduling guarantees, i.e. no timing guarantees, only functional correctness. It proves that the code doesn't access dangling pointers, implements the algorithm in the spec, does not break its invariants (as far as stated), things like that. We're essentially discussing what we think the spec *should* be saying in this bit so that we can be confident enough the timing properties actually hold and can be proved at a later point.

This specific invariant is mostly concerned with proving that the refill list is always ordered, and the code tries to establish an "obvious" reason for the invariant to remain true (a reason that is stronger than strictly necessary, but easier to track). We were actually hoping your intuition would say whether this interferes with the scheduling guarantees or not. We think it's fine, but we have not proved it. 

Are you worried about a specific aspect of the guarantees?

(You could argue that functional correctness of a scheduling mechanism should include its timing properties, and we do want to include them in the future, but the current proof framework doesn't really deal with time on the system level, so that has to be a later step.)

Reply:
Much of the rest of the implementation relies on this invariant being true:

 * we assume a task is ready if its head refill is in the past under the assumption that the head refill is the earliest refill, and
 * we assume that we can merge to refills by delaying one refill until it becomes adjacent to the subsequent refill in the list under the assumption that the subsequent refill is chronologically after it.

Reply:
After thinking about this a bit, I don't see a problem with the guarantees. It's "unusual" (from a RT textbook perspective), or at least somewhat different from how I have previously implemented this in another context, but it should not interfere with the isolation guarantees.